### PR TITLE
Improve how ConnectionAcquisitionTimeout is checked

### PR DESCRIPTION
**Pull request by ali-ince**

### CODE REVIEW

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs [hunk 1]
```diff
@@ -174,118 +169,160 @@ public IConnection Acquire(AccessMode mode)
...
                     }
-                    ThrowObjectDisposedException();
-                }
 
-                return connection;
-            });
+                    return connection;
+                });
+            }
+            catch (OperationCanceledException ex)
```
Comment (zhenlineo) [inline Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs]: I am thinking that this `try-catch` should be inside `return TryExecute`. TryExecute is only used to log errors before throwing the error to user. I would think this `OperationCanceledException` is less interesting to an end user than `ClientException`

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs [hunk 2]
```diff
@@ -174,118 +169,160 @@ public IConnection Acquire(AccessMode mode)
...
-                            break;
+                            do
+                            {
+                                if (!IsConnectionPoolFull())
+                                {
+                                    connection = CreateNewPooledConnection();
+                                    break;
+                                }
+
+                                if (_availableConnections.TryTake(out connection, 0, cancellationToken))
```
Comment (zhenlineo) [inline Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs]: Why removed `ConnAcquisitionQueuingTimeout`? We initially want this blocking queue is to make thread wait a bit on this `TryTake` method to introduce some queuing/waiting inside async. I am a bit worried if we return directly without any sleep, will the thread span here a lot?

**Review by zhenlineo — COMMENTED** [review]
Totally agree with the CancelationToken conn acquisition part, but I am a bit concerned to totally remove `ConnAcquisitionQueuingTimeout`.

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs (hunk 2 above)
Comment (ali-ince) [reply Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs]: I've put a small time lapse back into the loop (not a magnitude of seconds but milliseconds) but remove those timeouts and cancellationtoken checking logic from the first TryTake - making the fast path do not bother with those.

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs [hunk 3]
```diff
@@ -174,118 +171,161 @@ public IConnection Acquire(AccessMode mode)
...
+
+                        if (!_connectionValidator.IsValid(connection))
+                        {
+                            DestoryConnection(connection);
                         }
-                        if (_availableConnections.TryTake(out connection, ConnAcquisitionQueuingTimeout))
+                        else
                         {
                             break;
+                            ;
```
Comment (lutovich) [inline Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs]: Is this `;` intentional?

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs [hunk 4]
```diff
@@ -174,118 +171,161 @@ public IConnection Acquire(AccessMode mode)
...
+                            ThrowObjectDisposedException();
+                        }
+
+                        if (!_availableConnections.TryTake(out connection))
+                        {
+                            do
+                            {
+                                if (!IsConnectionPoolFull())
+                                {
+                                    connection = CreateNewPooledConnection();
```
Comment (lutovich) [inline Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs:203]: Unrelated to this PR: given a pool with `maxConnectionPoolSize - 1` acquired and `0` idle connections, can two threads that call `Acquire()` concurrently make pool expand to `maxConnectionPoolSize + 1` active connections? They will both try to create concurrently and attempts are probably not synchronized.

**Review by lutovich — COMMENTED** [review]
@ali-ince looks good, two small comments

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs (hunk 3 above)
Comment (ali-ince) [reply Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs]: No, indeed :)

**Comment by ali-ince** [comment]
I think I've addressed the synchronization issue you've pointed out.

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs (hunk 4 above)
Comment (zhenlineo) [reply Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs:203]: Except for correctness, I am not fan to strictly not exceed maxConnPoolSize. As it not only slows the code down also bring more complete logic.

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs [hunk 5]
```diff
@@ -151,6 +149,17 @@ public IPooledConnection CreateNewPooledConnection()
         private PooledConnection NewPooledConnection()
         {
             Interlocked.Increment(ref _poolSize);
+            if (_maxPoolSize != Config.Infinite && _poolSize > _maxPoolSize)
+            {
+                lock (_syncObject)
+                {
+                    if (_poolSize > _maxPoolSize)
+                    {
+                        throw new InvalidOperationException();
```
Comment (lutovich) [inline Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs]: Should we also decrement `_poolSize` here?

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs [hunk 6]
```diff
@@ -151,6 +149,17 @@ public IPooledConnection CreateNewPooledConnection()
         private PooledConnection NewPooledConnection()
         {
             Interlocked.Increment(ref _poolSize);
+            if (_maxPoolSize != Config.Infinite && _poolSize > _maxPoolSize)
```
Comment (lutovich) [inline Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs]: If we have two threads and `_poolSize == _maxPoolSize  - 1` they both will increment `_poolSize` and throw, right? I think one of them should succeed. We could probably use result of `Interlocked.Increment` to figure out which one should fail.

**Review by lutovich — COMMENTED** [review]
@ali-ince two more comments about synchronization around `maxPoolSize`

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs (hunk 4 above)
Comment (zhenlineo) [reply Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs:203]: Suggesting to have the fix in another PR. Right now the scope of this PR is good enough and the problem is not critical to block 1.5.

**Pull request merged**